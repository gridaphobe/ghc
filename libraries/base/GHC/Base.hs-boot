{-# LANGUAGE NoImplicitPrelude #-}

module GHC.Base where

import GHC.Types

(++) :: [a] -> [a] -> [a]

type String = [Char]
