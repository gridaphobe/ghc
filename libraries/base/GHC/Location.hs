{-# LANGUAGE RecordWildCards #-}
module GHC.Location
  ( CallStack
  , getCallStack

  , SrcLoc
  , srcLocPackage
  , srcLocModule
  , srcLocFile
  , srcLocStartLine
  , srcLocStartCol
  , srcLocEndLine
  , srcLocEndCol

  -- * Pretty printing
  , showCallStack
  , showSrcLoc
  ) where

--------------------------------------------------------------------------------
-- Source Locations
--------------------------------------------------------------------------------

-- | A CallStack is comprised of at least one 'SrcLoc'.
data CallStack = CallStack { getCallStack :: [SrcLoc] }
  deriving (Show, Eq)

showCallStack :: CallStack -> String
showCallStack (CallStack (root:rest))
  = unlines (showSrcLoc root : map (indent . showSrcLoc) rest)
  where
  indent l = "  " ++ l
showCallStack _ = error "CallStack cannot be empty!"

-- | A single location in the source code.
data SrcLoc = SrcLoc
  { srcLocPackage   :: String
  , srcLocModule    :: String
  , srcLocFile      :: String
  , srcLocStartLine :: Int
  , srcLocStartCol  :: Int
  , srcLocEndLine   :: Int
  , srcLocEndCol    :: Int
  } deriving (Show, Eq)

showSrcLoc :: SrcLoc -> String
showSrcLoc SrcLoc {..}
  = concat [ srcLocFile, ":"
           , show srcLocStartLine, ":"
           , show srcLocStartCol, " in "
           , srcLocPackage, ":", srcLocModule
           ]
