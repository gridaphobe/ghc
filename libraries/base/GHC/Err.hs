{-# LANGUAGE Trustworthy #-}
{-# LANGUAGE NoImplicitPrelude, MagicHash #-}
{-# OPTIONS_HADDOCK hide #-}

-----------------------------------------------------------------------------
-- |
-- Module      :  GHC.Err
-- Copyright   :  (c) The University of Glasgow, 1994-2002
-- License     :  see libraries/base/LICENSE
--
-- Maintainer  :  cvs-ghc@haskell.org
-- Stability   :  internal
-- Portability :  non-portable (GHC extensions)
--
-- The "GHC.Err" module defines the code for the wired-in error functions,
-- which have a special type in the compiler (with \"open tyvars\").
--
-- We cannot define these functions in a module where they might be used
-- (e.g., "GHC.Base"), because the magical wired-in type will get confused
-- with what the typechecker figures out.
--
-----------------------------------------------------------------------------

module GHC.Err( error ) where
import GHC.CString ()
import GHC.Types
import GHC.Prim
import GHC.Integer ()   -- Make sure Integer is compiled first
                        -- because GHC depends on it in a wired-in way
                        -- so the build system doesn't see the dependency
import {-# SOURCE #-} GHC.Exception( errorCallException )

-- | 'error' stops execution and displays an error message.
error :: [Char] -> CallStack -> a
error s stk = raise# (errorCallException s stk)
