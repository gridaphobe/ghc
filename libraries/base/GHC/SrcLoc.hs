{-# LANGUAGE NoImplicitPrelude, RecordWildCards #-}

-- | @since 4.8.2.0
module GHC.SrcLoc
  ( SrcLoc
  , srcLocPackage
  , srcLocModule
  , srcLocFile
  , srcLocStartLine
  , srcLocStartCol
  , srcLocEndLine
  , srcLocEndCol

  -- * Pretty printing
  , showSrcLoc
  ) where

import GHC.Base
import GHC.Show

-- | Pretty print 'SrcLoc'
--
-- @since 4.8.2.0
showSrcLoc :: SrcLoc -> [Char]
showSrcLoc SrcLoc {..}
  = foldr (++) ""
      [ srcLocFile, ":"
      , show srcLocStartLine, ":"
      , show srcLocStartCol, " in "
      , srcLocPackage, ":", srcLocModule
      ]

