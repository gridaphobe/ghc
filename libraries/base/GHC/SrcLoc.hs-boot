{-# LANGUAGE Trustworthy #-}
{-# LANGUAGE NoImplicitPrelude #-}

module GHC.SrcLoc
  ( SrcLoc
  , srcLocPackage
  , srcLocModule
  , srcLocFile
  , srcLocStartLine
  , srcLocStartCol
  , srcLocEndLine
  , srcLocEndCol
  ) where

import GHC.Types( Char, Int )

-- | A single location in the source code.
data SrcLoc = SrcLoc
  { srcLocPackage   :: [Char]
  , srcLocModule    :: [Char]
  , srcLocFile      :: [Char]
  , srcLocStartLine :: Int
  , srcLocStartCol  :: Int
  , srcLocEndLine   :: Int
  , srcLocEndCol    :: Int
  }
