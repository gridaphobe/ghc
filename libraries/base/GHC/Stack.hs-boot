-- This SOURCE-imported hs-boot module cuts a dependency loop:
--
--          GHC.Stack
-- imports  GHC.Base
-- imports  GHC.Err
-- imports  {-# SOURCE #-} GHC.Stack

{-# LANGUAGE Trustworthy #-}
{-# LANGUAGE NoImplicitPrelude #-}

module GHC.Stack ( CallStack, showCallStack ) where

import {-# SOURCE #-} GHC.SrcLoc ( SrcLoc )
import GHC.Types ( Char )

data CallStack = CallStack { getCallStack :: [([Char], SrcLoc)] }

showCallStack :: CallStack -> [Char]
