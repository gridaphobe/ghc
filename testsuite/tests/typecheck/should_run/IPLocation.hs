{-# LANGUAGE ImplicitParams, RankNTypes #-}
{-# OPTIONS_GHC -dcore-lint #-}
module Main where

import GHC.Exception
import GHC.Types

f0 = putStrLn $ showCallStack ?loc
     -- should be empty

f1 :: (?loc :: CallStack) => IO ()
f1 = putStrLn $ showCallStack ?loc
     -- should show the location of f1's call-site

f3 :: ((?loc :: CallStack) => () -> IO ()) -> IO ()
f3 x = x ()
       -- the call-site for the functional argument should be added to the
       -- stack..

f4 :: (?loc :: CallStack) => ((?loc :: CallStack) => () -> IO ()) -> IO ()
f4 x = x ()
       -- as should the call-site for f4 itself

f5 :: (?loc1 :: CallStack) => ((?loc2 :: CallStack) => () -> IO ()) -> IO ()
f5 x = x ()
       -- we only push new call-sites onto CallStacks with the name IP name

f6 :: (?loc :: CallStack) => Int -> IO ()
f6 0 = putStrLn $ showCallStack ?loc
f6 n = f6 (n-1)
       -- recursive functions add a SrcLoc for each recursive call

main = do f0
          f1
          f3 (\ () -> putStrLn $ showCallStack ?loc)
          f4 (\ () -> putStrLn $ showCallStack ?loc)
          f5 (\ () -> putStrLn $ showCallStack ?loc3)
          f6 5
