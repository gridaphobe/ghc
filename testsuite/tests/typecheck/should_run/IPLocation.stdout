<<<<<<< 896dd5f745660636f6c624bd5f18d67a5296229e
CallStack (from ImplicitParams):
  ?loc, called at IPLocation.hs:8:31 in main:Main
CallStack (from ImplicitParams):
  ?loc, called at IPLocation.hs:12:31 in main:Main
  f1, called at IPLocation.hs:40:11 in main:Main
CallStack (from ImplicitParams):
  ?loc, called at IPLocation.hs:16:34 in main:Main
  f2, called at IPLocation.hs:41:11 in main:Main
CallStack (from ImplicitParams):
  ?loc, called at IPLocation.hs:17:34 in main:Main
  f2, called at IPLocation.hs:41:11 in main:Main
CallStack (from ImplicitParams):
  ?loc, called at IPLocation.hs:42:48 in main:Main
  x, called at IPLocation.hs:22:8 in main:Main
CallStack (from ImplicitParams):
  ?loc, called at IPLocation.hs:43:48 in main:Main
  x, called at IPLocation.hs:27:8 in main:Main
  f4, called at IPLocation.hs:43:11 in main:Main
CallStack (from ImplicitParams):
  ?loc3, called at IPLocation.hs:44:48 in main:Main
CallStack (from ImplicitParams):
  ?loc, called at IPLocation.hs:35:33 in main:Main
  f6, called at IPLocation.hs:36:8 in main:Main
  f6, called at IPLocation.hs:36:8 in main:Main
  f6, called at IPLocation.hs:36:8 in main:Main
  f6, called at IPLocation.hs:36:8 in main:Main
  f6, called at IPLocation.hs:36:8 in main:Main
  f6, called at IPLocation.hs:45:11 in main:Main
=======
CallStack: empty
CallStack:
  f1, called at IPLocation.hs:34:11 in main:Main
CallStack:
  x, called at IPLocation.hs:16:8 in main:Main
CallStack:
  x, called at IPLocation.hs:21:8 in main:Main
  f4, called at IPLocation.hs:36:11 in main:Main
CallStack: empty
CallStack:
  f6, called at IPLocation.hs:30:8 in main:Main
  f6, called at IPLocation.hs:30:8 in main:Main
  f6, called at IPLocation.hs:30:8 in main:Main
  f6, called at IPLocation.hs:30:8 in main:Main
  f6, called at IPLocation.hs:30:8 in main:Main
  f6, called at IPLocation.hs:38:11 in main:Main
>>>>>>> revise callstack solver to
